Added product's customer view button on adminhtml product edit page

marcin-dykas opened the pull request:
<!--- Provide a general summary of the Pull Request in the Title above -->

### Description
<!--- Provide a description of the changes proposed in the pull request -->

This feature for Magento 2 adds additional button in admin panel product view. 

Button leads to frontend product page.
Allows user to see product being configured easily.

Button is disabled when product is not salable.

Saves time for administrators while working with catalog containing several products. Administrator can quickly see product in frontend without searching or creating URL manually.

### Fixed Issues (if relevant)
<!--- Provide a list of fixed issues in the format magento/magento2#<issue_number>, if relevant  -->
None - new additional function

### Manual testing scenarios
<!--- Provide a set of unambiguous steps to test the proposed code change -->
1. Go to admin panel
2. Go to CATALOG -> Products
3. Edit product
4. On edit product page new button is visible [Customer View] Button is between Back button Add Attribute button
5. After clicking product is being opened in new tab
6. If product is out of stock or just not salable button is disabled

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

Commit:
Added product's customer view button on adminhtml product edit page

magento-cicd2 commented:
[![CLA assistant check](https://cla.dev.magento.com/pull/badge/not_signed)](https://cla.dev.magento.com/magento/magento2?pullRequest=12078) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dev.magento.com/magento/magento2?pullRequest=12078) before we can accept your contribution.<br/>

Commit:
Fixed

FILE: .../code/Magento/Catalog/Block/Adminhtml/Product/Edit/Button/CustomerView.php
--------------------------------------------------------------------------------
FOUND 1 ERROR AFFECTING 1 LINE
--------------------------------------------------------------------------------
 36 | ERROR | [x] The closing parenthesis and the opening brace of a multi-line
    |       |     function declaration must be on the same line

dmanners commented:
Thanks @Hypo386 for this PR. My main question to you about this feature is how would it work for multi-store? Which frontend will it pick up in the end?

Hypo386 commented:
Hi @dmanners 

Well since registry('current_store'); 

is being used as parameters to 

\Magento\Framework\UrlInterface::getUrl()

it will produce URL for each store

e.g.
/catalog/product/view/id/1/?___store=en

or for default store view 
/catalog/product/view/id/1/?___store=default


Functionality is being done based on existing M2 function in CMS module where you can get URL for pages. Why can't we have this for products ?


The most tricky part here was that all URL "produced" in admin panel contains admin section like /admin_r4qkvf//catalog/product/view/id/1/ which was not acceptable. I had to include urlinterface from frontend to get frontend URLs.

dmanners commented:
Ok I see. The problem with that is that it is not recommended to use the registry or global state in general. It might be a nice idea to allow the admin user to choose the store they would like to view the product in as you could have a multi-store setup and this would also allow for the admin user to preview different product data if you had store specific data.

I do think this feature would be great to have so I have reached out to the product owner inside Magento to try and get some feedback on this for you.

Commit:
Changed registry('current_store') to storeManager->getStore()

Hypo386 commented:
You are right. I've changed registry('current_store') to storeManager->getStore()

I don't know if I understood correctly about stores because my idea was to use standard storeView switch. When you switch admin panel product form is being reloaded. Button will have different URL depends on what has been chosen.

![image](https://user-images.githubusercontent.com/4815371/33208633-967649b4-d112-11e7-94c7-af262a0822c9.png)

Depends on that has been chosen button will lead to 
/catalog/product/view/id/1/?___store=default
or
/catalog/product/view/id/1/?___store=pl
or
/catalog/product/view/id/1/?___store=en

dmanners commented:
@Hypo386 so which ever store view I am looking at in the admin is the store view I will preview? I like that idea nice.

dmanners commented on app/code/Magento/Catalog/Block/Adminhtml/Product/Edit/Action/UrlBuilder.php:
```diff
@@ -0,0 +1,55 @@
+    /**
+     * @var \Magento\Framework\UrlInterface
+     */
+    protected $frontendUrlBuilder;
```
by default new class properties can be set to private.

dmanners commented on app/code/Magento/Catalog/Block/Adminhtml/Product/Edit/Button/CustomerView.php:
```diff
@@ -0,0 +1,89 @@
+    /**
+     * @return string
+     */
+    public function getCustomerViewUrl()
```
If this method is only used in this class then I would suggest making it private.

dmanners commented on app/code/Magento/Catalog/Block/Adminhtml/Product/Edit/Action/UrlBuilder.php:
```diff
@@ -0,0 +1,55 @@
+    public function getUrl($routePath, $product, $scope, $store)
+    {
+        $this->frontendUrlBuilder->setScope($scope);
+        $href = $this->frontendUrlBuilder->getUrl(
```
You could simply return here rather than setting to `$href`

dmanners commented on app/code/Magento/Catalog/Block/Adminhtml/Product/Edit/Action/UrlBuilder.php:
```diff
@@ -0,0 +1,55 @@
+     * @param string $store
+     * @return string
+     */
+    public function getUrl($routePath, $product, $scope, $store)
```
If `$product` needs the `getId` method then I would suggest adding a type hint here to enforce the type of object passed in. In this case I would suggest the `\Magento\Catalog\Api\Data\ProductInterface` as this will enforce the `getId` method implementation.

dmanners commented on app/code/Magento/Catalog/Block/Adminhtml/Product/Edit/Button/CustomerView.php:
```diff
@@ -0,0 +1,89 @@
+class CustomerView extends CoreGeneric
+{
+    /** @var \Magento\Catalog\Model\Product */
+    protected $product;
```
Again these can all be private

dmanners reviewed (changes requested):
We can close the scope of some properties and methods also would you be able to add some tests to cover new classes?

Commit:
Changed visibility scope after CR

Hypo386 commented:
Yes idea was simple. What storeview you see is what (URL) you get. Simple as that.

Thx for precise CR. I've changed visibility scope in code.

Now I will try to provide some tests

dmanners committed:
Clean up some unnecessary variables and wrong typehints

dmanners commented:
@Hypo386 looks good, I added 1 commit to clean up a few typehints and variables but it looks good. Let me know if you need some help with the tests.

Commit:
Added first unit test

Hypo386 commented:
@dmanners yes please

I've added unit test. It's not working. 

I don't know what should I do with CustomerView::getCustomerViewUrl()

I've been trying to mockup actionUrlBuilder to set result of \Magento\Catalog\Block\Adminhtml\Product\Edit\Action\UrlBuilder::getUrl function but I'm not sure is it correct.

Should I try to mockup this function like in my commit or should I mockup UrlBuilder instance to mockup getUrl for UrlBuilder?

If UrlBuilder should be mockedup it should be done by getMockBuilder or by ObjectManager? 

UrlBuilder is another file and I don't know is it's correct way to have it in test for CustomerView

dmanners committed:
Add basic getButtonData test case

dmanners commented:
@Hypo386 I have updated the test so that it will now run. The only real change I made was to add the returns for mocked methods.

```
        $storeMock = $this->getMockBuilder(\Magento\Store\Model\Store::class)
            ->disableOriginalConstructor()
            ->getMock();
        $storeManagerMock->expects($this->any())
            ->method('getStore')
            ->willReturn($storeMock);
        $productMock = $this->getMockBuilder(ProductInterface::class)
            ->setMethods(['getId', 'isSalable'])
            ->getMockForAbstractClass();
        $productMock->expects($this->once())
            ->method('isSalable')
            ->willReturn(true);
        $productMock->expects($this->once())
            ->method('getId')
            ->willReturn(12);
```

Also I then simplified the action mock as follows.

```
$actionUrlBuilderMock->expects($this->once())
            ->method('getUrl')
            ->willReturn('test_url');
```

With regards to the test case I think this is fine but it would be good to add a couple of more tests for the differences in:

```
if (!$product->isSalable() || !$product->getId()) {
            $buttonData['disabled'] = 'disabled';
        }
```

Would you be able to add these in for me? Also fee free to set it back up with a `setup` method and the properties. I just found it a bit easier for debugging without them.

Commit:
Fixed bugs in unitTest

Hypo386 commented:
I have just fixed this as well but you have been faster

I'm starting to understand what is going on in tests here so I think I can try....

dmanners commented:
@Hypo386 cool, as always if you need help just drop me a message.

Commit:
Refactor for unit test.

Added more tests for disabled product

Commit:
Refactor for unit test.

Added more tests for disabled product

Commit:
Added test file to check namespaces

Commit:
Removed not necessary file

Commit:
Added different separator lines

Hypo386 commented:
During Travis tests there was strange error:

```
1) Magento\Test\Integrity\ClassesTest::testClassNamespaces
Passed: 19078, Failed: 1, Incomplete: 0, Skipped: 0.
Data set: /home/travis/build/magento/magento2/app/code/Magento/Catalog/Test/Unit/Block/Adminhtml/Product/Edit/Button/CustomerViewTest.php
Location of /home/travis/build/magento/magento2/app/code/Magento/Catalog/Test/Unit/Block/Adminhtml/Product/Edit/Button/CustomerViewTest.php does not match formal namespace: Magento\Catalog\Test\Unit\Block\Adminhtml\Product\Edit\Button

2) Magento\Test\Integrity\ClassesTest::testClassReferences
Passed: 19078, Failed: 1, Incomplete: 0, Skipped: 0.
Data set: /home/travis/build/magento/magento2/app/code/Magento/Catalog/Test/Unit/Block/Adminhtml/Product/Edit/Button/CustomerViewTest.php
Incorrect namespace usage(s) found in file /home/travis/build/magento/magento2/app/code/Magento/Catalog/Test/Unit/Block/Adminhtml/Product/Edit/Button/CustomerViewTest.php:
```





I've been debuging ClassesTest.php by 

`dev/tests/static$ /usr/bin/php7.0 /var/www/html/magento2contribute/./vendor/phpunit/phpunit/phpunit -vvv testsuite/Magento/Test/Integrity/ClassesTest.php`

There is assertEquals which doesn't want to pass two variables

string(64) "Magento/Catalog/Block/Adminhtml/Product/Edit/Button/CustomerView"
string(64) "Magento/Catalog/Block/Adminhtml/Product/Edit/Button/CustomerView"

I've found that sth is wrong with unvisible signs there in those strings. 

Test passes only when '\r' sign is line separator in files 

app/code/Magento/Catalog/Block/Adminhtml/Product/Edit/Button/CustomerView.php
app/code/Magento/Catalog/Test/Unit/Block/Adminhtml/Product/Edit/Button/CustomerViewTest.php

At least here locally 

> PHPUnit 6.2.4 by Sebastian Bergmann and contributors.
> 
> Runtime:       PHP 7.0.19-1 with Xdebug 2.5.0
> Configuration: /var/www/html/magento2contribute/dev/tests/static/phpunit.xml.dist
> 
> ......                                                              6 / 6 (100%)
> 
> Time: 17.64 seconds, Memory: 40.00MB
> 

I guess app/code/Magento/Catalog/Block/Adminhtml/Product/Edit/Button/CustomerView.php has been modified using Mac. I've pulled it by git. Additionally there is always difference in PHPUnit version.

Is it possible ? :)

dmanners commented:
@Hypo386 interesting, I will have a look at this. It could have been me when working on the initial unit test. Let me get back to you on this.

Hypo386 commented:
Now it doesn't make sense because other test

  1 | ERROR | [x] End of line character is invalid; expected "\n" but found
    |       |     "\r"

Commit:
Added different separator lines \n

Commit:
Fixed problems with namespaces

dmanners commented:
@Hypo386 I notice that Travis is has now successfully run. Is there anything else you are waiting for or do you feel this PR is now ready?

Hypo386 commented:
@dmanners 

I think this is it.

What are next steps ?

dmanners commented:
@Hypo386 I will process this PR at the moment. If you would like to pick up other issues in new PRs then yes either find the issue on the board or TODO label you would like to work on and comment that you would like to work on it. Then one of the team should be able to assign the issue to you.

dmanners commented:
@Hypo386 thank you for your pull request. Unfortunately as it currently stands we are not able to take this feature into Magento for the following reasons.

As it stands this just creates a front end url. It would be best if it was a Preview of changed made by the admin or "Show me this product as customer X" so you could see this product as a customer in group X with all custom prices and information,

This could be a good idea to take further as a community driven project or even an extension of the staging system. Feel free to reach out to the team if you would be interested in taking this on with the help of our PO and development team.

https://github.com/magento/community-features/issues is a good place to add new feature ideas/requests.
